ICE: `missing binding mode`

<!--
[31mICE[0m: Rustc ./a.rs '' 'thread 'rustc' panicked at compiler/rustc_hir_typeck/src/upvar.rs:1791:70: 'missing binding mode'', 'thread 'rustc' panicked at compiler/rustc_hir_typeck/src/upvar.rs:1791:70: 'missing binding mode''
File: /tmp/im/a.rs
-->

auto-reduced (treereduce-rust):
````rust
async unsafe extern "cmse-nonsecure-entry" fn multiple_named_lifetimes<T: Unpin>(_: u8, ...) {}

````

original:
````rust
#![feature("{}" a)]

async unsafe extern "C-cmse-nonsecure-entry" fn multiple_named_lifetimes<T: Unpin>(_: u8, ...) { while main { } }
//~^ ERROR hidden type for `impl Future<Output = ()>` captures lifetime that does not appear in bounds

fn my_fn(_args: &[A]) {
  println!("hello world");
}

````

Version information
````
rustc 1.84.0-nightly (97ae1df8a 2024-10-25)
binary: rustc
commit-hash: 97ae1df8aa9fa6dfd29dced18b232d25208c4111
commit-date: 2024-10-25
host: x86_64-unknown-linux-gnu
release: 1.84.0-nightly
LLVM version: 19.1.1

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/97ae1df8aa9fa6dfd29dced18b232d25208c4111/compiler/rustc_hir_typeck/src/upvar.rs#L1785-L1797

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0670]: `async fn` is not permitted in Rust 2015
 --> /tmp/icemaker_global_tempdir.4Ax9UanUShFv/rustc_testrunner_tmpdir_reporting.IEMd1DzRNcyR/mvce.rs:1:1
  |
1 | async unsafe extern "C-cmse-nonsecure-entry" fn multiple_named_lifetimes<T: Unpin>(_: u8, ...) {}
  | ^^^^^ to use `async fn`, switch to Rust 2018 or later
  |
  = help: pass `--edition 2021` to `rustc`
  = note: for more on editions, read https://doc.rust-lang.org/edition-guide

error: only foreign, `unsafe extern "C"`, or `unsafe extern "C-unwind"` functions may have a C-variadic arg
 --> /tmp/icemaker_global_tempdir.4Ax9UanUShFv/rustc_testrunner_tmpdir_reporting.IEMd1DzRNcyR/mvce.rs:1:91
  |
1 | async unsafe extern "C-cmse-nonsecure-entry" fn multiple_named_lifetimes<T: Unpin>(_: u8, ...) {}
  |                                                                                           ^^^

error[E0658]: C-cmse-nonsecure-entry ABI is experimental and subject to change
 --> /tmp/icemaker_global_tempdir.4Ax9UanUShFv/rustc_testrunner_tmpdir_reporting.IEMd1DzRNcyR/mvce.rs:1:21
  |
1 | async unsafe extern "C-cmse-nonsecure-entry" fn multiple_named_lifetimes<T: Unpin>(_: u8, ...) {}
  |                     ^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #75835 <https://github.com/rust-lang/rust/issues/75835> for more information
  = help: add `#![feature(cmse_nonsecure_entry)]` to the crate attributes to enable
  = note: this compiler was built on 2024-10-25; consider upgrading it if it is out of date

error[E0658]: C-variadic functions are unstable
 --> /tmp/icemaker_global_tempdir.4Ax9UanUShFv/rustc_testrunner_tmpdir_reporting.IEMd1DzRNcyR/mvce.rs:1:1
  |
1 | async unsafe extern "C-cmse-nonsecure-entry" fn multiple_named_lifetimes<T: Unpin>(_: u8, ...) {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #44930 <https://github.com/rust-lang/rust/issues/44930> for more information
  = help: add `#![feature(c_variadic)]` to the crate attributes to enable
  = note: this compiler was built on 2024-10-25; consider upgrading it if it is out of date

error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.4Ax9UanUShFv/rustc_testrunner_tmpdir_reporting.IEMd1DzRNcyR/mvce.rs:1:98
  |
1 | async unsafe extern "C-cmse-nonsecure-entry" fn multiple_named_lifetimes<T: Unpin>(_: u8, ...) {}
  |                                                                                                  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.4Ax9UanUShFv/rustc_testrunner_tmpdir_reporting.IEMd1DzRNcyR/mvce.rs`

thread 'rustc' panicked at compiler/rustc_hir_typeck/src/upvar.rs:1791:70:
missing binding mode
stack backtrace:
   0:     0x77075a09d83a - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::hcedb2fcda5cdd07c
   1:     0x77075a8041ca - core::fmt::write::haa85bbf9373b67e8
   2:     0x77075baa39d1 - std::io::Write::write_fmt::hd17b6cdad4d5e705
   3:     0x77075a09d692 - std::sys::backtrace::BacktraceLock::print::h50f59795d6a1f7e3
   4:     0x77075a09fb96 - std::panicking::default_hook::{{closure}}::h0966ca81a9d29c3a
   5:     0x77075a09f9e0 - std::panicking::default_hook::h534e5141a1b02198
   6:     0x770759119f9f - std[a151dedc4facde64]::panicking::update_hook::<alloc[946be05e1a1394f6]::boxed::Box<rustc_driver_impl[bf30803df350f481]::install_ice_hook::{closure#0}>>::{closure#0}
   7:     0x77075a0a02a8 - std::panicking::rust_panic_with_hook::h0090c4c2e6e5d265
   8:     0x77075a0a007a - std::panicking::begin_panic_handler::{{closure}}::h5174eb866483fbac
   9:     0x77075a09dce9 - std::sys::backtrace::__rust_end_short_backtrace::h334590b8864eb5e5
  10:     0x77075a09fd3c - rust_begin_unwind
  11:     0x770756ae3320 - core::panicking::panic_fmt::h307156d14c26e2e0
  12:     0x77075713fbab - core::option::expect_failed::heb168b2a112ea99d
  13:     0x77075c174c06 - <rustc_hir_typeck[a8c3eaf7852a1958]::fn_ctxt::FnCtxt>::determine_capture_mutability.cold
  14:     0x77075b2352d0 - <rustc_hir_typeck[a8c3eaf7852a1958]::fn_ctxt::FnCtxt>::compute_min_captures
  15:     0x770757910477 - <rustc_hir_typeck[a8c3eaf7852a1958]::fn_ctxt::FnCtxt>::analyze_closure
  16:     0x7707575d564a - <rustc_hir_typeck[a8c3eaf7852a1958]::upvar::InferBorrowKindVisitor as rustc_hir[227b0424fb17eaca]::intravisit::Visitor>::visit_expr
  17:     0x77075aaea4a5 - rustc_hir_typeck[a8c3eaf7852a1958]::typeck
  18:     0x77075aae9ac7 - rustc_query_impl[6ffb465cf6366f2]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[6ffb465cf6366f2]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 8usize]>>
  19:     0x77075ac43201 - rustc_query_system[e2e7e315caf8fc79]::query::plumbing::try_execute_query::<rustc_query_impl[6ffb465cf6366f2]::DynamicConfig<rustc_query_system[e2e7e315caf8fc79]::query::caches::VecCache<rustc_span[916daa99866cf1f4]::def_id::LocalDefId, rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[6ffb465cf6366f2]::plumbing::QueryCtxt, false>
  20:     0x77075ac41857 - rustc_query_impl[6ffb465cf6366f2]::query_impl::typeck::get_query_non_incr::__rust_end_short_backtrace
  21:     0x77075b178e0a - rustc_middle[58c8be1544d4ff96]::query::plumbing::query_get_at::<rustc_query_system[e2e7e315caf8fc79]::query::caches::VecCache<rustc_span[916daa99866cf1f4]::def_id::LocalDefId, rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 8usize]>>>
  22:     0x77075b8ddbab - rustc_hir_analysis[f8c34c1d50f2c7d9]::collect::type_of::type_of_opaque
  23:     0x77075b8ddae5 - rustc_query_impl[6ffb465cf6366f2]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[6ffb465cf6366f2]::query_impl::type_of_opaque::dynamic_query::{closure#2}::{closure#0}, rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 8usize]>>
  24:     0x77075ab854fa - rustc_query_system[e2e7e315caf8fc79]::query::plumbing::try_execute_query::<rustc_query_impl[6ffb465cf6366f2]::DynamicConfig<rustc_query_system[e2e7e315caf8fc79]::query::caches::DefIdCache<rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[6ffb465cf6366f2]::plumbing::QueryCtxt, false>
  25:     0x77075b996ff6 - rustc_query_impl[6ffb465cf6366f2]::query_impl::type_of_opaque::get_query_non_incr::__rust_end_short_backtrace
  26:     0x77075adef540 - rustc_middle[58c8be1544d4ff96]::query::plumbing::query_get_at::<rustc_query_system[e2e7e315caf8fc79]::query::caches::DefIdCache<rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 8usize]>>>
  27:     0x770757c2c334 - rustc_hir_analysis[f8c34c1d50f2c7d9]::collect::type_of::type_of
  28:     0x77075ab868aa - rustc_query_impl[6ffb465cf6366f2]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[6ffb465cf6366f2]::query_impl::type_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 8usize]>>
  29:     0x77075ab854fa - rustc_query_system[e2e7e315caf8fc79]::query::plumbing::try_execute_query::<rustc_query_impl[6ffb465cf6366f2]::DynamicConfig<rustc_query_system[e2e7e315caf8fc79]::query::caches::DefIdCache<rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[6ffb465cf6366f2]::plumbing::QueryCtxt, false>
  30:     0x77075ab850a3 - rustc_query_impl[6ffb465cf6366f2]::query_impl::type_of::get_query_non_incr::__rust_end_short_backtrace
  31:     0x77075b1296d0 - rustc_middle[58c8be1544d4ff96]::query::plumbing::query_get_at::<rustc_query_system[e2e7e315caf8fc79]::query::caches::DefIdCache<rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 8usize]>>>
  32:     0x77075aea761d - <rustc_trait_selection[45dc482989e69e08]::traits::query::normalize::QueryNormalizer as rustc_type_ir[543408bf699dc]::fold::FallibleTypeFolder<rustc_middle[58c8be1544d4ff96]::ty::context::TyCtxt>>::try_fold_ty
  33:     0x77075aea2ac8 - <rustc_traits[98b13475d2e688c2]::normalize_erasing_regions::provide::{closure#0} as core[301db181db02ff04]::ops::function::FnOnce<(rustc_middle[58c8be1544d4ff96]::ty::context::TyCtxt, rustc_middle[58c8be1544d4ff96]::ty::ParamEnvAnd<rustc_middle[58c8be1544d4ff96]::ty::generic_args::GenericArg>)>>::call_once
  34:     0x77075aea26f7 - rustc_query_impl[6ffb465cf6366f2]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[6ffb465cf6366f2]::query_impl::try_normalize_generic_arg_after_erasing_regions::dynamic_query::{closure#2}::{closure#0}, rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 8usize]>>
  35:     0x77075aea1e19 - rustc_query_system[e2e7e315caf8fc79]::query::plumbing::try_execute_query::<rustc_query_impl[6ffb465cf6366f2]::DynamicConfig<rustc_query_system[e2e7e315caf8fc79]::query::caches::DefaultCache<rustc_middle[58c8be1544d4ff96]::ty::ParamEnvAnd<rustc_middle[58c8be1544d4ff96]::ty::generic_args::GenericArg>, rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[6ffb465cf6366f2]::plumbing::QueryCtxt, false>
  36:     0x77075aea1b60 - rustc_query_impl[6ffb465cf6366f2]::query_impl::try_normalize_generic_arg_after_erasing_regions::get_query_non_incr::__rust_end_short_backtrace
  37:     0x77075ae9dc8b - <rustc_middle[58c8be1544d4ff96]::ty::normalize_erasing_regions::TryNormalizeAfterErasingRegionsFolder as rustc_type_ir[543408bf699dc]::fold::FallibleTypeFolder<rustc_middle[58c8be1544d4ff96]::ty::context::TyCtxt>>::try_fold_ty
  38:     0x77075afc8388 - rustc_ty_utils[3110184492a391f3]::layout::layout_of
  39:     0x77075afc7f19 - rustc_query_impl[6ffb465cf6366f2]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[6ffb465cf6366f2]::query_impl::layout_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 16usize]>>
  40:     0x77075afc71c8 - rustc_query_system[e2e7e315caf8fc79]::query::plumbing::try_execute_query::<rustc_query_impl[6ffb465cf6366f2]::DynamicConfig<rustc_query_system[e2e7e315caf8fc79]::query::caches::DefaultCache<rustc_middle[58c8be1544d4ff96]::ty::ParamEnvAnd<rustc_middle[58c8be1544d4ff96]::ty::Ty>, rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 16usize]>>, false, true, false>, rustc_query_impl[6ffb465cf6366f2]::plumbing::QueryCtxt, false>
  41:     0x77075afc6e6d - rustc_query_impl[6ffb465cf6366f2]::query_impl::layout_of::get_query_non_incr::__rust_end_short_backtrace
  42:     0x77075b8e66b2 - rustc_middle[58c8be1544d4ff96]::query::plumbing::query_get_at::<rustc_query_system[e2e7e315caf8fc79]::query::caches::DefaultCache<rustc_middle[58c8be1544d4ff96]::ty::ParamEnvAnd<rustc_middle[58c8be1544d4ff96]::ty::Ty>, rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 16usize]>>>
  43:     0x7707592f168c - rustc_hir_analysis[f8c34c1d50f2c7d9]::hir_ty_lowering::cmse::is_valid_cmse_output
  44:     0x77075b171ce6 - <dyn rustc_hir_analysis[f8c34c1d50f2c7d9]::hir_ty_lowering::HirTyLowerer>::lower_fn_ty
  45:     0x77075b16ffe4 - rustc_hir_analysis[f8c34c1d50f2c7d9]::collect::infer_return_ty_for_fn_sig
  46:     0x77075b172555 - rustc_hir_analysis[f8c34c1d50f2c7d9]::collect::fn_sig
  47:     0x77075aa5be6b - rustc_query_impl[6ffb465cf6366f2]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[6ffb465cf6366f2]::query_impl::fn_sig::dynamic_query::{closure#2}::{closure#0}, rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 24usize]>>
  48:     0x77075aa5f064 - rustc_query_system[e2e7e315caf8fc79]::query::plumbing::try_execute_query::<rustc_query_impl[6ffb465cf6366f2]::DynamicConfig<rustc_query_system[e2e7e315caf8fc79]::query::caches::DefIdCache<rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[6ffb465cf6366f2]::plumbing::QueryCtxt, false>
  49:     0x77075aa5e8e2 - rustc_query_impl[6ffb465cf6366f2]::query_impl::fn_sig::get_query_non_incr::__rust_end_short_backtrace
  50:     0x77075ab7adbe - <rustc_hir_analysis[f8c34c1d50f2c7d9]::collect::CollectItemTypesVisitor as rustc_hir[227b0424fb17eaca]::intravisit::Visitor>::visit_item
  51:     0x7707577bc428 - rustc_hir_analysis[f8c34c1d50f2c7d9]::check::wfcheck::check_well_formed
  52:     0x77075ac04197 - rustc_query_impl[6ffb465cf6366f2]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[6ffb465cf6366f2]::query_impl::check_well_formed::dynamic_query::{closure#2}::{closure#0}, rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 1usize]>>
  53:     0x77075ac038e7 - rustc_query_system[e2e7e315caf8fc79]::query::plumbing::try_execute_query::<rustc_query_impl[6ffb465cf6366f2]::DynamicConfig<rustc_query_system[e2e7e315caf8fc79]::query::caches::VecCache<rustc_span[916daa99866cf1f4]::def_id::LocalDefId, rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[6ffb465cf6366f2]::plumbing::QueryCtxt, false>
  54:     0x77075ac03550 - rustc_query_impl[6ffb465cf6366f2]::query_impl::check_well_formed::get_query_non_incr::__rust_end_short_backtrace
  55:     0x77075ac04410 - rustc_hir_analysis[f8c34c1d50f2c7d9]::check::wfcheck::check_mod_type_wf
  56:     0x77075ac04247 - rustc_query_impl[6ffb465cf6366f2]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[6ffb465cf6366f2]::query_impl::check_mod_type_wf::dynamic_query::{closure#2}::{closure#0}, rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 1usize]>>
  57:     0x77075b3d5745 - rustc_query_system[e2e7e315caf8fc79]::query::plumbing::try_execute_query::<rustc_query_impl[6ffb465cf6366f2]::DynamicConfig<rustc_query_system[e2e7e315caf8fc79]::query::caches::DefaultCache<rustc_span[916daa99866cf1f4]::def_id::LocalModDefId, rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[6ffb465cf6366f2]::plumbing::QueryCtxt, false>
  58:     0x77075b3d54f3 - rustc_query_impl[6ffb465cf6366f2]::query_impl::check_mod_type_wf::get_query_non_incr::__rust_end_short_backtrace
  59:     0x77075ac3effb - rustc_hir_analysis[f8c34c1d50f2c7d9]::check_crate
  60:     0x77075ac33517 - rustc_interface[2b41161da54db00e]::passes::run_required_analyses
  61:     0x77075b54925e - rustc_interface[2b41161da54db00e]::passes::analysis
  62:     0x77075b549231 - rustc_query_impl[6ffb465cf6366f2]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[6ffb465cf6366f2]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 1usize]>>
  63:     0x77075b73816e - rustc_query_system[e2e7e315caf8fc79]::query::plumbing::try_execute_query::<rustc_query_impl[6ffb465cf6366f2]::DynamicConfig<rustc_query_system[e2e7e315caf8fc79]::query::caches::SingleCache<rustc_middle[58c8be1544d4ff96]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[6ffb465cf6366f2]::plumbing::QueryCtxt, false>
  64:     0x77075b737e4f - rustc_query_impl[6ffb465cf6366f2]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  65:     0x77075b6045b1 - rustc_interface[2b41161da54db00e]::interface::run_compiler::<core[301db181db02ff04]::result::Result<(), rustc_span[916daa99866cf1f4]::ErrorGuaranteed>, rustc_driver_impl[bf30803df350f481]::run_compiler::{closure#0}>::{closure#1}
  66:     0x77075b6a3454 - std[a151dedc4facde64]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[2b41161da54db00e]::util::run_in_thread_with_globals<rustc_interface[2b41161da54db00e]::util::run_in_thread_pool_with_globals<rustc_interface[2b41161da54db00e]::interface::run_compiler<core[301db181db02ff04]::result::Result<(), rustc_span[916daa99866cf1f4]::ErrorGuaranteed>, rustc_driver_impl[bf30803df350f481]::run_compiler::{closure#0}>::{closure#1}, core[301db181db02ff04]::result::Result<(), rustc_span[916daa99866cf1f4]::ErrorGuaranteed>>::{closure#0}, core[301db181db02ff04]::result::Result<(), rustc_span[916daa99866cf1f4]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[301db181db02ff04]::result::Result<(), rustc_span[916daa99866cf1f4]::ErrorGuaranteed>>
  67:     0x77075b6a388d - <<std[a151dedc4facde64]::thread::Builder>::spawn_unchecked_<rustc_interface[2b41161da54db00e]::util::run_in_thread_with_globals<rustc_interface[2b41161da54db00e]::util::run_in_thread_pool_with_globals<rustc_interface[2b41161da54db00e]::interface::run_compiler<core[301db181db02ff04]::result::Result<(), rustc_span[916daa99866cf1f4]::ErrorGuaranteed>, rustc_driver_impl[bf30803df350f481]::run_compiler::{closure#0}>::{closure#1}, core[301db181db02ff04]::result::Result<(), rustc_span[916daa99866cf1f4]::ErrorGuaranteed>>::{closure#0}, core[301db181db02ff04]::result::Result<(), rustc_span[916daa99866cf1f4]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[301db181db02ff04]::result::Result<(), rustc_span[916daa99866cf1f4]::ErrorGuaranteed>>::{closure#1} as core[301db181db02ff04]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  68:     0x77075b6a432b - std::sys::pal::unix::thread::Thread::new::thread_start::h254586afceceda1d
  69:     0x7707558a339d - <unknown>
  70:     0x77075592849c - <unknown>
  71:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.84.0-nightly (97ae1df8a 2024-10-25) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `multiple_named_lifetimes`
#1 [type_of_opaque] computing type of opaque `multiple_named_lifetimes::{opaque#0}`
end of query stack
error: aborting due to 5 previous errors

Some errors have detailed explanations: E0601, E0658, E0670.
For more information about an error, try `rustc --explain E0601`.

```

</p>
</details>

<!--
query stack:
#0 [typeck] type-checking `multiple_named_lifetimes`
#1 [type_of_opaque] computing type of opaque `multiple_named_lifetimes::{opaque#0}`

-->

--- a/compiler/rustc_hir_analysis/src/hir_ty_lowering/cmse.rs
+++ b/compiler/rustc_hir_analysis/src/hir_ty_lowering/cmse.rs
@@ -85,6 +85,12 @@ pub(crate) fn validate_cmse_abi<'tcx>(
                 return;
             };
 
+            // An `extern "cmse-nonsecure-entry"` function cannot be c-variadic. We run
+            // into https://github.com/rust-lang/rust/issues/132142 if we don't explicitly bail.
+            if decl.c_variadic {
+                return;
+            }
+
             match is_valid_cmse_inputs(tcx, fn_sig) {
                 Ok(Ok(())) => {}
                 Ok(Err(index)) => {

--- a/tests/crashes/132142.rs
+++ b/tests/crashes/132142.rs
@@ -1,3 +1,0 @@
-//@ known-bug: #132142
-
-async extern "cmse-nonsecure-entry" fn fun(...) {}

--- a/tests/ui/cmse-nonsecure/cmse-nonsecure-entry/c-variadic.rs
+++ b/tests/ui/cmse-nonsecure/cmse-nonsecure-entry/c-variadic.rs
@@ -1,0 +1,69 @@
+//@ add-core-stubs
+//@ edition: 2018
+//@ compile-flags: --target thumbv8m.main-none-eabi --crate-type lib
+//@ needs-llvm-components: arm
+#![feature(cmse_nonsecure_entry, c_variadic, no_core, lang_items)]
+#![no_core]
+
+extern crate minicore;
+use minicore::*;
+
+#[lang = "va_list"]
+struct VaList(*mut u8);
+
+unsafe extern "cmse-nonsecure-entry" fn c_variadic(_: u32, _: ...) {
+    //~^ ERROR `...` is not supported for `extern "cmse-nonsecure-entry"` functions
+}
+
+// A regression test for https://github.com/rust-lang/rust/issues/132142
+async unsafe extern "cmse-nonsecure-entry" fn async_and_c_variadic(_: ...) {
+    //~^ ERROR `...` is not supported for `extern "cmse-nonsecure-entry"` functions
+    //~| ERROR functions cannot be both `async` and C-variadic
+}
+
+// Below are the lang items that are required for a program that defines an `async` function.
+// Without them, the ICE that is tested for here is not reached for this target. For now they are in
+// this file, but they may be moved into `minicore` if/when other `#[no_core]` tests want to use
+// them.
+
+// NOTE: in `core` this type uses `NonNull`.
+#[lang = "ResumeTy"]
+pub struct ResumeTy(*mut Context<'static>);
+
+#[lang = "future_trait"]
+pub trait Future {
+    /// The type of value produced on completion.
+    #[lang = "future_output"]
+    type Output;
+
+    // NOTE: misses the `poll` method.
+}
+
+#[lang = "async_drop"]
+pub trait AsyncDrop {
+    // NOTE: misses the `drop` method.
+}
+
+#[lang = "Poll"]
+pub enum Poll<T> {
+    #[lang = "Ready"]
+    Ready(T),
+
+    #[lang = "Pending"]
+    Pending,
+}
+
+#[lang = "Context"]
+pub struct Context<'a> {
+    // NOTE: misses a bunch of fields.
+    _marker: PhantomData<fn(&'a ()) -> &'a ()>,
+}
+
+#[lang = "get_context"]
+pub unsafe fn get_context<'a, 'b>(cx: ResumeTy) -> &'a mut Context<'b> {
+    // NOTE: the actual implementation looks different.
+    mem::transmute(cx.0)
+}
+
+#[lang = "pin"]
+pub struct Pin<T>(T);

--- a/tests/ui/cmse-nonsecure/cmse-nonsecure-entry/c-variadic.stderr
+++ b/tests/ui/cmse-nonsecure/cmse-nonsecure-entry/c-variadic.stderr
@@ -1,0 +1,28 @@
+error: `...` is not supported for `extern "cmse-nonsecure-entry"` functions
+  --> $DIR/c-variadic.rs:14:60
+   |
+LL | unsafe extern "cmse-nonsecure-entry" fn c_variadic(_: u32, _: ...) {
+   |        -----------------------------                       ^^^^^^
+   |        |
+   |        `extern "cmse-nonsecure-entry"` because of this
+   |
+   = help: only `extern "C"` and `extern "C-unwind"` functions may have a C variable argument list
+
+error: functions cannot be both `async` and C-variadic
+  --> $DIR/c-variadic.rs:19:1
+   |
+LL | async unsafe extern "cmse-nonsecure-entry" fn async_and_c_variadic(_: ...) {
+   | ^^^^^ `async` because of this                                      ^^^^^^ C-variadic because of this
+
+error: `...` is not supported for `extern "cmse-nonsecure-entry"` functions
+  --> $DIR/c-variadic.rs:19:68
+   |
+LL | async unsafe extern "cmse-nonsecure-entry" fn async_and_c_variadic(_: ...) {
+   |              -----------------------------                         ^^^^^^
+   |              |
+   |              `extern "cmse-nonsecure-entry"` because of this
+   |
+   = help: only `extern "C"` and `extern "C-unwind"` functions may have a C variable argument list
+
+error: aborting due to 3 previous errors
+

--- a/tests/ui/cmse-nonsecure/cmse-nonsecure-entry/generics.rs
+++ b/tests/ui/cmse-nonsecure/cmse-nonsecure-entry/generics.rs
@@ -1,7 +1,7 @@
 //@ add-core-stubs
 //@ compile-flags: --target thumbv8m.main-none-eabi --crate-type lib
 //@ needs-llvm-components: arm
-#![feature(cmse_nonsecure_entry, c_variadic, no_core, lang_items)]
+#![feature(cmse_nonsecure_entry, no_core, lang_items)]
 #![no_core]
 
 extern crate minicore;
@@ -64,9 +64,4 @@ struct WrapperTransparent<'a>(&'a dyn Trait);
 extern "cmse-nonsecure-entry" fn wrapped_trait_object(x: WrapperTransparent) -> WrapperTransparent {
     //~^ ERROR return value of `"cmse-nonsecure-entry"` function too large to pass via registers [E0798]
     x
-}
-
-unsafe extern "cmse-nonsecure-entry" fn c_variadic(_: u32, _: ...) {
-    //~^ ERROR `...` is not supported for `extern "cmse-nonsecure-entry"` functions
-    //~| ERROR requires `va_list` lang_item
 }

--- a/tests/ui/cmse-nonsecure/cmse-nonsecure-entry/generics.stderr
+++ b/tests/ui/cmse-nonsecure/cmse-nonsecure-entry/generics.stderr
@@ -1,13 +1,3 @@
-error: `...` is not supported for `extern "cmse-nonsecure-entry"` functions
-  --> $DIR/generics.rs:69:60
-   |
-LL | unsafe extern "cmse-nonsecure-entry" fn c_variadic(_: u32, _: ...) {
-   |        -----------------------------                       ^^^^^^
-   |        |
-   |        `extern "cmse-nonsecure-entry"` because of this
-   |
-   = help: only `extern "C"` and `extern "C-unwind"` functions may have a C variable argument list
-
 error[E0798]: functions with the `"cmse-nonsecure-entry"` ABI cannot contain generics in their type
   --> $DIR/generics.rs:30:1
    |
@@ -71,12 +61,6 @@ LL | extern "cmse-nonsecure-entry" fn wrapped_trait_object(x: WrapperTransparent
    = note: functions with the `"cmse-nonsecure-entry"` ABI must pass their result via the available return registers
    = note: the result must either be a (transparently wrapped) i64, u64 or f64, or be at most 4 bytes in size
 
-error: requires `va_list` lang_item
-  --> $DIR/generics.rs:69:60
-   |
-LL | unsafe extern "cmse-nonsecure-entry" fn c_variadic(_: u32, _: ...) {
-   |                                                            ^^^^^^
-
-error: aborting due to 9 previous errors
+error: aborting due to 7 previous errors
 
 For more information about this error, try `rustc --explain E0798`.

